aten/src/ATen/native/vulkan/VulkanAten.cpp

```diff
@@ -272,6 +272,18 @@ Tensor vulkan_add(const Tensor& self, const Tensor& other, const Scalar alpha) {
   return new_with_vtensor_vulkan(std::move(output), self.options());
 }
 
+Tensor& vulkan_add_(Tensor& self, const Tensor& other, Scalar alpha) {
+  auto& x = vtensor_from_vulkan(self.is_vulkan() ? self : self.vulkan());
+  auto& y = vtensor_from_vulkan(other.is_vulkan() ? other : other.vulkan());
+  float a = alpha.to<float>();
+
+  VulkanTensor output{self.sizes().vec()};
+  output.allocate_storage();
+  vulkan::detail::add(output, x, y, a);
+  x = std::move(output);
+  return self;
```

Comment:
I'm not sure I understand this logic here.  If self is already allocated and the goal of this function is to implement c += a * b, then do we need to allocate output as well?  Can we not do it in place in 'self'?  Also, why the move?  Can't we just pass output, output, and input and a (in that order) to vulkan::detail::add?

Reply:
Here we read from x (self) and y (other) and write output values to output.  
At the moment all shaders work having separate output and input(s), here it is using the same shader that is used for copy variant of this op.

`x = std::move(output)` replaces x (self) content (pointer to VulkanTensor::Impl) of x to the result of output.

`vulkan::detail::add()` uses only `image()` representation, so we do not really need allocated buffer(). 
For this implementation we need `output.allocate_storage()`, but with changes  https://github.com/pytorch/pytorch/pull/42569 allocate_storage can be removed